UserAutocomplete+CoreDataClass.swift

```diff
@@ -0,0 +1,7 @@
+import Foundation
+import CoreData
+
+@objc(UserAutocomplete)
+public class UserAutocomplete: NSManagedObject {
```

Comment:
I guess that even though this entity exists for the autocomplete, in the end, it's just usernames.

Then I was looking into our database schema and we already have a very similar entity which is `BlogAuthor`. All the username suggestions are users added to a specific blog, right?

Maybe that's something you have already tried but `BlogAuthor` wouldn't suit us here?

Reply:
Good question, I hadn't seen `BlogAuthor`. Since `Suggestion` already existed, I didn't think to look for another class that could handle this — thanks for bringing it up.

One thing I didn't mention in this PR is that the Gutenberg team is currently implementing x-posting, which will be very similar to @-mentions.

Now the question comes up of whether instead of `UserAutocomplete` I could reuse `BlogAuthor` — and while I'm at it, use `Blog` in the future for x-posting. There's a clear advantage to reusing the entities we already have instead of creating new ones, especially if they make sense semantically. On the other hand, some things to keep in mind are:
- to audit existing use of `BlogAuthor` and `Blog` to make sure that existing screens will not be affected by the fact that these same entities will now be used for additional functionality (mentions and x-posting)
- both entities have non-optional attributes that mentions and x-posts won't be able to use (e.g. `userID` on `BlogAuthor`) and many non-optional properties on `Blog` that the x-post endpoint doesn't return — converting to non-optional could require significant effort

I think a good idea would be to check if the backend services that provide data for mentions and x-posts internally use the same entities as the services that populate `BlogAuthor` and `Blog` objects. If they do, that would be a good indication that we should do the same, but if not, that we should keep the separate entities. I'll try to look into this a bit and see what I can find, but please let me know if you have an thoughts on the above.

Reply:
I agree with all your points. It would be awesome if we could reuse those entities (since, correct me if I'm wrong, the mentions are blog authors) but we need to be cautious about changing what already exists.

I checked the payload for the suggestions and they have: `ID`, `user_login`, `display_name` and `image_URL` (it's missing `email`,`linkedUserID` and `primaryBlogID`).

I know that `BasePost` is used by `Post` and `ReaderPost`, for example. Maybe instead of using `BlogAuthor` we can just do a similar thing? Then this won't affect the parts of the code that already uses it.

Regarding `x-posts` I don't think `Blog` will suit us in this case.

Reply:
> I checked the payload for the suggestions and they have: ID, user_login, display_name and image_URL (it's missing email,linkedUserID and primaryBlogID).

`BlogAuthor` is populated by the `sites/$site_id/users` endpoint, which doesn't actually include the user's display name.
<details>
<summary>See JSON:</summary>
<pre>
    {
      "ID": 190912407,
      "login": "paulvonschrottky",
      "email": "paulvonschrottky@gmail.com",
      "name": "paulvonschrottky",
      "first_name": "",
      "last_name": "",
      "nice_name": "paulvonschrottky",
      "URL": "",
      "avatar_URL": "https://0.gravatar.com/avatar/f2976ad632f6923d61ebe9acbb8d852f?s=96&d=identicon&r=G",
      "profile_URL": "https://en.gravatar.com/paulvonschrottky",
      "ip_address": "",
      "site_ID": 171152168,
      "site_visible": true,
      "roles": [
        "editor"
      ],
      "is_super_admin": false
    },
</pre>
</details>

> I know that BasePost is used by Post and ReaderPost, for example. Maybe instead of using BlogAuthor we can just do a similar thing? Then this won't affect the parts of the code that already uses it.

To do this properly, it would probably be best to remove `displayName` from `BlogAuthor` (since it's not currently used and doesn't belong there) and put it on the new `BlogAuthor` subclass. 

> Regarding x-posts I don't think Blog will suit us in this case.
Yeah, that's my feeling as well.

Reply:
(fyi: I've made significant changes since the previous comment)

Instead of using two entities as I did before, I now just use one new entity (`AtMentionSuggestion`). By adding a 1-to-many nullable relationship from `Blog` to `AtMentionSuggestion`, I haven't lost any of the functionality of the previous approach while making things much simpler.
My gut feeling is that this is the most intuitive approach (rather than involving `BlogAuthor`) and only adds one new entity. 
(A potential downside is: imagine a user with multiple sites, where each site has hundreds of @-mention suggestions (think a8c). Even though the author list is probably much the same for each site, we store each site's list of @-mentions separately. E.g. a user who has 5 sites each with 1000 @-mentions suggestions would result in 5000 rows in the `AtMentionSuggestion` entity table. The upside is that this is cleared when the user closes their session.)